infrastructure/templates/public-api/main.bicep

```diff
@@ -116,68 +118,137 @@ module fileShareModule 'components/fileShares.bicep' = {
   }
 }
 
+// In order to link PostgreSQL Flexible Server to a VNet, it must have a Private DNS zone available with a name ending
+// with "postgres.database.azure.com".
+resource postgreSqlPrivateDnsZone 'Microsoft.Network/privateDnsZones@2020-06-01' = {
+  name: 'private.postgres.database.azure.com'
+  location: 'global'
+  resource vNetLink 'virtualNetworkLinks' = {
+    name: '${resourcePrefix}-psql-flexibleserver-vnet-link'
+    location: 'global'
+    properties: {
+      registrationEnabled: false
+      virtualNetwork: {
+        id: vNetModule.outputs.vNetRef
+      }
+    }
+  }
+}
+
 // Deploy PostgreSQL Database.
 module postgreSqlServerModule 'components/postgresqlDatabase.bicep' = {
   name: 'postgreSQLDatabaseDeploy'
   params: {
     resourcePrefix: resourcePrefix
     location: location
+    createMode: psqlDbUsersAdded ? 'Update' : 'Default'
     adminName: postgreSqlAdminName!
     adminPassword: postgreSqlAdminPassword!
     dbSkuName: postgreSqlSkuName
     dbStorageSizeGB: postgreSqlStorageSizeGB
     dbAutoGrowStatus: postgreSqlAutoGrowStatus
-    keyVaultName: keyVaultName
+    postgreSqlVersion: '16'
     tagValues: tagValues
-    vNetId: vNetModule.outputs.vNetRef
+    privateDnsZoneId: postgreSqlPrivateDnsZone.id
     firewallRules: postgreSqlFirewallRules
     subnetId: vNetModule.outputs.postgreSqlSubnetRef
     databaseNames: ['public_data']
   }
-  dependsOn: [
-    vNetModule
-  ]
+}
+
+resource apiContainerAppManagedIdentity 'Microsoft.ManagedIdentity/userAssignedIdentities@2023-01-31' = {
+  name: apiContainerAppManagedIdentityName
+  location: location
+}
+
+var acrPullRole = resourceId('Microsoft.Authorization/roleDefinitions', '7f951dda-4ed3-4680-a7ca-43fe172d538d')
+
+@description('This allows the managed identity of the container app to access the registry, note scope is applied to the wider ResourceGroup not the ACR')
+resource apiContainerAppManagedIdentityRBAC 'Microsoft.Authorization/roleAssignments@2022-04-01' = {
+  name: guid(resourceGroup().id, apiContainerAppManagedIdentity.id, acrPullRole)
+  properties: {
+    roleDefinitionId: acrPullRole
+    principalId: apiContainerAppManagedIdentity.properties.principalId
+    principalType: 'ServicePrincipal'
+  }
+}
```

Comment:
I think I might have to trial running this on the Dev Resource Group, and make the necessary amends there should we run into issues.  I'm not sure either way what we'll have provisioned in terms of permissions for our existing service connections but we did a chunk of work with them a while back when rolling out the PNPM work for the first time.

Reply:
Ok yep I agree about trialling it out to find out.

The PNPM work is different because that's an app service running a container, not an app within a container app environment. That work also opted to use username/password authentication rather than a user managed identity where the identity would be granted the AcrPull role.

I can't see that the AcrPull role has any assignments for the existing ACR or that it's even a recommended way of doing it. We should look into sorting it at some point